### PR TITLE
[BE] Migrate some tests to pytest

### PR DESCRIPTION
**Pull request by ebsmothers**
Summary: Migrated FLAVA, two tower, late fusion, and CNNLSTM tests from unittest to pytest.
Kept initialization order the same so we don't have to change any expected values in this PR (most relevant for FLAVA tests).

Test plan:
```
python -m pytest -v
```

### CODE REVIEW

File: test/models/flava/test_flava.py [hunk 1]
```diff
@@ -155,50 +188,84 @@ def setUp(self):
...
             multimodal=TransformerOutput(),
-            text=self.text_encoder(
+            text=text_encoder(
                 text, return_attn_weights=True, return_hidden_states=True
             ),
-            image=self.image_encoder(image),
-            image_masked=self.image_encoder(image),
+            image=image_encoder(image),
+            image_masked=image_encoder(image),
             projected_image_embeddings=actual.image.last_hidden_state[:, 0, :],
```
Comment (langong347) [inline test/models/flava/test_flava.py:252]: `expected` depends on `actual`? !

```
projected_image_embeddings=actual.image.last_hidden_state[:, 0, :],
projected_text_embeddings=actual.text.last_hidden_state[:, 0, :],
```

Generally, it'd be good not to use components we test to calculate expected even if the values cannot be computed by hand to ensure isolation. For example, `text`, `image` and `image_masked` fields are computed from the low level components which are part of flava model. 

I'd suggest at least remove the explicit dependency on `actual` for the last two fields (and check if there're similar cases elsewhere). Your call for those that are derived from low level components.

File: test/models/flava/test_flava.py [hunk 2]
```diff
@@ -209,43 +276,44 @@ def test_forward_masked_image_and_text(self):
...
             multimodal=TransformerOutput(),
-            text=self.text_encoder(
+            text=text_encoder(
                 text, return_attn_weights=True, return_hidden_states=True
             ),
-            image=self.image_encoder(image),
-            image_masked=self.image_encoder(image, masked_image),
+            image=image_encoder(image),
+            image_masked=image_encoder(image, masked_image),
             projected_image_embeddings=actual.image.last_hidden_state[:, 0, :],
```
Comment (langong347) [inline test/models/flava/test_flava.py:284]: remove `expected` dep on `actual`

**Review by langong347 — COMMENTED** [review]
Thanks for migrating those tests!

File: test/models/flava/test_flava.py [hunk 3]
```diff
@@ -72,18 +99,19 @@ def test_forward_pretraining(self):
...
             mlm_labels=mlm_labels,
         )
-        self.assertIsNone(output.mlm_output)
-        self.assertIsNone(output.mim_output)
-        self.assertIsNotNone(output.global_contrastive_output)
-        self.assertIsNotNone(output.mmm_text_output)
-        self.assertIsNotNone(output.mmm_image_output)
-        self.assertIsNotNone(output.itm_output)
-        self.assertAlmostEqual(
+        assert output.mlm_output is None
```
Comment (langong347) [inline test/models/flava/test_flava.py:102]:
```
expected = {
    "mlm_output": None,
    "mim_output": None,
    "global_constrastive_output": (size, sum),
    "losses": (size, 21.5150),
}
assert_expected_namedtuple(actual, expected, rtol=1e-5, atol=1e-4)
```

File: test/models/flava/test_flava.py (hunk 1 above)
Comment (ebsmothers) [reply test/models/flava/test_flava.py:252]: Yeah this is a good point. Let me derive these separately without relying on `actual`

File: test/models/flava/test_flava.py [hunk 4]
```diff
@@ -22,44 +22,71 @@
...
+            text,
+            image,
+            image_for_codebook,
+            image_patches_mask,
+            text_masked,
+            mlm_labels,
+            itm_labels,
+        )
+
+    @torch.no_grad()
```
[the remark this answers is not in the document]
Comment (ebsmothers) [reply test/models/flava/test_flava.py:60]: In this case it may actually be useful to keep it; the pretraining test is using the full FLAVA model architecture and is one of our slower tests. I will remove the `no_grad` decorator and test locally as a sanity check though

**Comment by ebsmothers** [comment]
> By order of initialization do you mean the order of the fixtures for each model layer? That shouldn't matter anymore if:

@langong347 I mean that for some existing tests multiple components are initialized with random weights without resetting the seed between (e.g. [here](https://github.com/facebookresearch/multimodal/blob/main/test/models/flava/test_flava.py#L31-L34)). To avoid remapping the expected values in this PR I initialize any such components in the same fixture to maintain the same seed status. Lmk if you have an alternative suggestion here though

File: test/models/flava/test_flava.py [hunk 5]
```diff
@@ -155,50 +188,84 @@ def setUp(self):
...
+        self,
+        image_encoder,
+        text_encoder,
+        mm_encoder,
+        image_to_mm_projection,
+        text_to_mm_projection,
+        text_projection,
+        image_projection,
+    ):
+        return FLAVAModel(
```
[the remark this answers is not in the document]
Comment (ebsmothers) [reply test/models/flava/test_flava.py:222]: I think the behavior is the same in both cases. FLAVA sets dropout prob to 0.0 and uses layer norm, which is unaffected by eval mode. So we could add it but I don't think it would actually change anything

File: test/models/flava/test_flava.py (hunk 3 above)
Comment (ebsmothers) [reply test/models/flava/test_flava.py:102]: Ultimately decided to leave this as is, mainly because the test is currently not explicitly checking sizes and values of all these components and the final loss value check should be sufficient here

File: test/models/flava/test_flava.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (langong347) [reply test/models/flava/test_flava.py:222]: I think as a good practice we should add it because the model may at some point introduce randomness. What do you think?

**Comment by langong347** [comment]
> I mean that for some existing tests multiple components are initialized with random weights without resetting the seed between (e.g. [here](https://github.com/facebookresearch/multimodal/blob/main/test/models/flava/test_flava.py#L31-L34)). To avoid remapping the expected values in this PR I initialize any such components in the same fixture to maintain the same seed status. Lmk if you have an alternative suggestion here though

Since the test data are defined in the test module, their order of initialization we have full visibilty.
My comment was more around the initialization of model layers where a fresh random seed in between the fixtures can be helpful to shield impact from trivial refactoring (e.g., reordering of model layers in the model class).

File: test/models/flava/test_flava.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ebsmothers) [reply test/models/flava/test_flava.py:222]: I will just add it so it doesn't lead to any confusion in the future

**Review by langong347 — APPROVED** [review]
